### PR TITLE
Cherrypicking allocation policy fixes to 1.10

### PR DESCRIPTION
**Pull request by VSadov**
Also added a new kind of OOM error - OutOfMemoryAllocatorPolicy_unrecoverable_error

### CODE REVIEW

File: lib/Runtime/Base/ThreadContext.h [hunk 1]
```diff
@@ -622,7 +622,7 @@ class ThreadContext sealed :
     bool reentrancySafeOrHandled;
     bool isInReentrancySafeRegion;
 
-    AllocationPolicyManager * allocationPolicyManager;
+    AllocationPolicyManager * allocationPolicyManager;
```
Comment (dilijev) [inline lib/Runtime/Base/ThreadContext.h]: nit: added space?

File: lib/Common/Memory/Recycler.h [hunk 2]
```diff
@@ -1146,6 +1146,8 @@ class Recycler
 #ifdef NTBUILD
     void SetTelemetryBlock(RecyclerWatsonTelemetryBlock * telemetryBlock) { this->telemetryBlock = telemetryBlock; }
 #endif
+    
+    uint GetPinnedObjectCount() const { return this->pinnedObjectMap.Count(); }
```
Comment (dilijev) [inline lib/Common/Memory/Recycler.h]: Is this used anywhere? Maybe in Full? Is there a corresponding Full review?

Comment (VSadov) [reply lib/Common/Memory/Recycler.h]: It is used in Full. I will make a PR once Full passes tests. 
These are existing changes though. They merged just after we forked 1.10, but we want them there.

**Comment by VSadov** [comment]
closing as dupe of #5297

**Pull request closed without merging**